### PR TITLE
support more complete implicit aggregation usage in openCypher

### PR DESCRIPTION
**Pull request by jievince**

### CODE REVIEW

File: src/visitor/AggregateVisitor.cpp [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+    found_ = found;
+}
+
+void AggregateVisitor::visit(MapExpression* expr) {
+    DEFINE_DEFER_LIST
+    bool found = false;
+
+    for (const auto& pair : expr->items()) {
+        pair.second->accept(this);
+        found = found || found_;
```
Comment (czpmango) [inline src/visitor/AggregateVisitor.cpp:109]: rename `found` or `found_`

File: src/visitor/AggregateVisitor.h [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+#ifndef VISITOR_AGGREGATEVISITOR_H_
+#define VISITOR_AGGREGATEVISITOR_H_
+
+#include "common/expression/Expression.h"
+#include "visitor/ExprVisitorImpl.h"
+
+namespace nebula {
+namespace graph {
+class AggregateVisitor final : public ExprVisitorImpl {
```
Comment (czpmango) [inline src/visitor/AggregateVisitor.h:15]: Is this visitor only for aggregation, or is there a more general abstraction to it in other rewrite scenarios?

File: src/visitor/RewriteVisitor.h [hunk 3]
```diff
@@ -36,7 +36,7 @@ class RewriteVisitor final : public ExprVisitorImpl {
         return rewriter_;
     }
 
-private:
+public:
```
Comment (czpmango) [inline src/visitor/RewriteVisitor.h:39]: How about keeping it private and implementing it in a static function？[https://github.com/vesoft-inc/nebula-graph/pull/1305](https://github.com/vesoft-inc/nebula-graph/pull/1305)

File: src/validator/MatchValidator.cpp [hunk 4]
```diff
@@ -694,44 +695,68 @@ Status MatchValidator::validateGroup(YieldClauseContext &yieldCtx) const {
...
-                         .ok()) {
-                    return Status::SemanticError("Aggregate function nesting is not allowed: `%s'",
-                                                 colExpr->toString().c_str());
+        auto aggs = ExpressionUtils::collectAll(colExpr, {Expression::Kind::kAggregate});
+        for (auto *agg : aggs) {
+            DCHECK_EQ(agg->kind(), Expression::Kind::kAggregate);
+            if (!ExpressionUtils::checkAggExpr(static_cast<const AggregateExpression *>(agg))
+                     .ok()) {
+                return Status::SemanticError("Aggregate function nesting is not allowed: `%s'",
+                                             colExpr->toString().c_str());
```
Comment (czpmango) [inline src/validator/MatchValidator.cpp:704]: NG_RETURN_IF_ERROR(ExpressionUtils::checkAggExpr(aggExpr));

File: src/validator/MatchValidator.cpp [hunk 5]
```diff
@@ -694,44 +695,68 @@ Status MatchValidator::validateGroup(YieldClauseContext &yieldCtx) const {
...
+            }
+        }
+    }
```
Comment (czpmango) [inline src/validator/MatchValidator.cpp:707]: Extract this code block into function (Maybe YieldClauseContext::checkAggregation).

File: src/validator/MatchValidator.cpp [hunk 6]
```diff
@@ -694,44 +695,68 @@ Status MatchValidator::validateGroup(YieldClauseContext &yieldCtx) const {
...
+
+    yieldCtx.needGenProject_ = true;
+    for (auto *col : cols) {
+        auto *colExpr = col->expr();
+        auto colOldName = col->toString();
+        auto foldRes = ExpressionUtils::foldConstantExpr(colExpr);
+        NG_RETURN_IF_ERROR(foldRes);
+        auto *newExpr = foldRes.value();
```
Comment (czpmango) [inline src/validator/MatchValidator.cpp:715]: You only need to do some check here(maybe use `ExpressionUtils::isEvaluableExpr`), it is better for the optimizer to do constant folding.

File: src/validator/MatchValidator.cpp [hunk 7]
```diff
@@ -694,44 +695,68 @@ Status MatchValidator::validateGroup(YieldClauseContext &yieldCtx) const {
...
+
+        if (newExpr->kind() == Expression::Kind::kAggregate) {
+            yieldCtx.groupItems_.emplace_back(newExpr);
+            yieldCtx.aggOutputColumnNames_.emplace_back(newExpr->toString());
+            yieldCtx.projCols_->addColumn(new YieldColumn(
+                VariablePropertyExpression::make(pool, "", colOldName), colOldName));
+            yieldCtx.projOutputColumnNames_.emplace_back(colOldName);
+        } else {
+            AggregateVisitor visitor;
```
Comment (czpmango) [inline src/validator/MatchValidator.cpp:724]: Merge this if-else block. You always set these variables in the if-block, but don't set `groupKeys_`, isn’t it?

Comment (czpmango) [reply src/validator/MatchValidator.cpp:724]:
```cpp
Status MatchValidator::validateGroup(YieldClauseContext &yieldCtx) const {
    NG_RETURN_IF_ERROR(YieldClauseContext::checkAggregation()); 
    YieldClauseContext::ExtractGroupItems(); // hiding the visitor's logic
   
    return Status::OK();
}

```

**Comment by jievince** [comment]
This pr will be unuseful if GQL standard do not support implicitly aggregation

**Pull request closed without merging**